### ISSUE TITLE
Vim mode multi-key mappings

### ISSUE THREAD

**Issue by cjfuller** [zed-industries/community#176]
**Is your feature request related to a problem? Please describe.**
I normally remap keys in vim or vim-mode in other editors so that I can use `jk` in insert mode to exit insert mode instead of escape, but I don't see any way to do this from the docs, and it did not work to try to hack it with:
```
    {
        "context": "Editor && vim_mode == insert && vim_operator == j",
        "bindings": {
            "k": "vim::NormalBefore"
        }
    }
``` 

**Describe the solution you'd like**
I'd like to somehow be able to bind j then k to Escape in insert mode.

**Describe alternatives you've considered**
I tried the config hack above to no avail, and I also considered retraining to ctrl-c, which is the default alternate binding, but I'd rather be able to use something on the home row.

**Mockup**
For example, I'd love to be able to write the following:
```
    {
        "context": "Editor && vim_mode == insert",
        "bindings": {
            "jk": "vim::NormalBefore"
        }
    }
```

**Comment by albertorestifo** [comment]
Vim mode without the ability to remap multi-key bindings is very limiting indeed.

Another use-case: In normal mode, I want to remap `gd` to the Go To Definition command.

**Comment by caliguIa** [comment]
@albertorestifo you may have already figured this out but you can achieve go to definition already with the following snippet:

`[
    {
        "context": "Editor && vim_operator == g",
        "bindings": {
            "d": "editor::GoToDefinition"
        }
    }
]`

**Comment by mauscoelho** [comment]
The `vim_operator == g` but when trying to do something like `vim_operator == [` it does not work, is it a special character? How would it translate?

Not working example:
```
  {
    "context": "Editor && vim_operator == [",
    "bindings": {
      "d": "editor::GoToPrevDiagnostic"      
    }
  }
```

**Comment by washanhanzi** [comment]
This worked for `jk`:

```
  {
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "j k": "vim::NormalBefore"
    }
  }
```

but if you want to type `j`, it's `jj`.

**Comment by mauscoelho** [comment]
For reference this works for me:

```
{
    "context": "Editor && vim_mode == normal && (vim_operator == none || vim_operator == n) && !VimWaiting",
    "bindings": {
      // put key-bindings here if you want them to work only in normal mode
      "g d": "editor::GoToDefinition",
      "g r": "editor::FindAllReferences",
      "[ d": "editor::GoToPrevDiagnostic",
      "] d": "editor::GoToDiagnostic"
    }
  }
  ```

**Comment by calebmeyer** [comment]
I added a bunch of keybinds, and they're all working (amazingly fast and well compared to vs code!). My only issue is what @washanhanzi said, I now have to double type `j`s.

<details>
<summary>My new keybinds!</summary>
<pre>
[
    {
        // normal & visual mode
        "context": "Editor && VimControl && !VimWaiting && !menu",
        "bindings": {}
    },
    {
        // normal mode
        "context": "Editor && vim_mode == normal && (vim_operator == none || vim_operator == n) && !VimWaiting",
        "bindings": {
            "g d": "editor::GoToDefinition",
            "space space": "command_palette::Toggle",
            "space b d": "pane::CloseActiveItem",
            "space b n": "pane::ActivateNextItem",
            "space b p": "pane::ActivatePrevItem",
            "space c l": "editor::ToggleComments",
            "space f s": "workspace::Save",
            "space f S": "workspace::SaveAll",
            "space j f": "project_symbols::Toggle", // not quite right, since it's full project
            "space p s": "TODO", // go to project
            "space q a": "pane::CloseAllItems",
            "space q o": "pane::CloseInactiveItems",
            "space q w": "workspace::CloseWindow",
            "space q q": "zed::Quit",
            "space s d": "pane::SplitDown",
            "space s l": "pane::SplitLeft",
            "space s r": "pane::SplitRight",
            "space s u": "pane::SplitUp",
            "space t i": "editor::ToggleInlayHints",
            "space t n": "workspace::NewTerminal",
            "space t t": "terminal_panel::ToggleFocus",
            "space w d": [
                "workspace::ActivatePaneInDirection",
                "Down"
            ],
            "space w l": [
                "workspace::ActivatePaneInDirection",
                "Left"
            ],
            "space w r": [
                "workspace::ActivatePaneInDirection",
                "Right"
            ],
            "space w u": [
                "workspace::ActivatePaneInDirection",
                "Up"
            ],
            "space w w": "workspace::ActivateNextPane",
        }
    },
    {
        // visual, visual line & visual block modes
        "context": "Editor && vim_mode == visual && !VimWaiting && !VimObject",
        "bindings": {},
    },
    {
        // insert mode
        "context": "Editor && vim_mode == insert",
        "bindings": {
            "j k": "vim::NormalBefore"
        }
    }
]
<pre>
</details>

**Comment by mauscoelho** [comment]
> "space j f": "project_symbols::Toggle", // not quite right, since it's full project

You can use `cmd-shift-O` for buffer symbols @calebmeyer. I could not find the action name for it as well

**Comment by JosephTLyons** [comment]
> > "space j f": "project_symbols::Toggle", // not quite right, since it's full project
> 
> You can use `cmd-shift-O` for buffer symbols @calebmeyer. I could not find the action name for it as well

We call this action `outline: toggle`.

**Comment by ConradIrwin** [comment]
@calebmeyer thanks for sharing these! 

I'm going to add outline::Toggle and project_symbols::Toggle to the default map. Vim doesn't use `g s` for anything useful ("sleep for count seconds??!") and doesn't use `g S` at all.

I've already begun repurposing the `g` namespace for zed specific stuff, so that would give us:

```
      "g d": "editor::GoToDefinition",  // in vim "go to local declaration" 
      "g shift-d": "editor::GoToTypeDefinition", // in vim "go to global declaration"
      "g .": "editor::ToggleCodeActions", // zed specific
      "g shift-a": "editor::FindAllReferences", // zed specific
      "g s": "outline::Toggle",  // zed specific
      "g shift-s": "project_symbols::Toggle",  // zed specific
      // ... 
      "c d": "editor::Rename", // zed specific
```

Also a quick note that in your file `"space f S"` needs to be spelt `"space f shift-s"` for zed's keyboard handler to understand it.

**Comment by ThePyroTF2** [comment]
In the preview version of Zed, doing something like this:
```json
    {
        "context": "Editor && vim_mode == insert",
        "bindings": {
            "j k": "vim::NormalBefore",
            "k j": "vim::NormalBefore"
        }
    },
```
Actually ends up typing `k` or `j` in the file before executing the command. So I have to get rid of the extra letter every time I use the keybind. Not fun.

**Comment by d1y** [comment]
> In the preview version of Zed, doing something like this:
> 
> ```json
>     {
>         "context": "Editor && vim_mode == insert",
>         "bindings": {
>             "j k": "vim::NormalBefore",
>             "k j": "vim::NormalBefore"
>         }
>     },
> ```
> 
> Actually ends up typing `k` or `j` in the file before executing the command. So I have to get rid of the extra letter every time I use the keybind. Not fun.

Same problem, in zed-preview
![zed-bug](https://github.com/zed-industries/community/assets/45585937/9f7ce6a1-ed10-4ed2-ba6f-a2d8597326ab)

**Comment by ConradIrwin** [comment]
Thanks for the report on Preview. I'll at least revert to the gpui1 behavior, but I wonder if I can figure out how to teach the input handler about pending keys so we can have our cake and eat it to.

**Comment by JosephTLyons** [comment]
This should be shipped in [v0.119.16-pre](https://zed.dev/releases/preview/0.119.16).

**Comment by ConradIrwin** [comment]
I've added support for `jk` in insert-mode in the latest nightly:
* You'll need to set up bindings like:
```json
[
  {
    "context": "Editor && vim_mode == insert",
    "bindings": {
      "j k": "vim::NormalBefore"
    }
  }
]
```

* Once you have that, typing `jk` in insert mode will act like `<esc>`
* If you type `j` and then wait 1000ms, it will type `j`
* If you type `j` and then any other key, it will type the `j` just before the next key.

This works similarly in normal mode, for example if you have:

```json
[
  {
    "bindings": {
      "shift-escape": null,
      "ctrl-enter": "menu::ShowContextMenu",
      ", w": "workspace::Save"
    }
  }
]
```

* If you type `,w` then it will save.
* If you type `,` and then wait 1000ms it will do the default (repeat the most recent `tf` command in reverse).
* If you type `,` and then another key, it will do the default just before the next key

With that done, I think we can close this out.

There is a related feature request to support mapping from one sequence of keys to another sequence of keys, but we should track that separately!
